pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/invertedindex/RealtimeLuceneIndexReaderRefreshThread.java

```diff
@@ -0,0 +1,133 @@
+          // one or more realtime lucene readers (one per column) belonging to the MutableSegment
+          // are added to the global queue managed by this thread. The thread that signals will
+          // grab this mutex and signal on the condition variable.
+          //
+          // This refresh thread will be woken up (and grab the mutex automatically as per the
+          // implementation of await) and check if the queue is non-empty. It will then proceed to
+          // poll the queue and refresh the realtime index readers for the polled segment.
+          //
+          // The mutex and condition-variable semantics take care of the scenario when on
+          // a given Pinot server, there is no realtime segment with text index enabled. In such
+          // cases, there is no need for this thread to wake up simply after every few seconds/minutes
+          // only to find that there is nothing to be refreshed. The thread should simply be
+          // off CPU until signalled specifically. This also covers the situation where initially
+          // there were few realtime segments of a table with text index. Later if they got
+          // moved to another server as part of rebalance, then again there is no need for this thread
+          // to do anything until some realtime segment is created with text index enabled.
+          _conditionVariable.await();
+        } catch (InterruptedException e) {
+          LOGGER.warn("Realtime lucene reader refresh thread got interrupted while waiting on condition variable: ", e);
+          Thread.currentThread().interrupt();
+        } finally {
+          _mutex.unlock();
+        }
+      }
+      RealtimeLuceneReadersForRealtimeSegment realtimeReadersForSegment = _luceneRealtimeReaders.poll();
+      if (realtimeReadersForSegment != null) {
+        String segmentName = realtimeReadersForSegment.getSegmentName();
+        realtimeReadersForSegment.getLock().lock();
+        if (!realtimeReadersForSegment.isSegmentDestroyed()) {
+          // If the segment already got destroyed before this task picked up the segment
+          // from queue, we don't have to do anything
+          try {
+            List<RealtimeLuceneTextIndexReader> realtimeLuceneReaders = realtimeReadersForSegment.getRealtimeLuceneReaders();
+            for (RealtimeLuceneTextIndexReader realtimeReader : realtimeLuceneReaders) {
+              SearcherManager searcherManager = realtimeReader.getSearcherManager();
+              try {
+                searcherManager.maybeRefresh();
+              } catch (Exception e) {
+                LOGGER.warn("Caught exception {} while refreshing realtime lucene reader for segment: {}", e, segmentName);
+              }
+            }
+            _luceneRealtimeReaders.offer(realtimeReadersForSegment);
+          } finally {
+            realtimeReadersForSegment.getLock().unlock();
+          }
+        }
```

Comment:
If we find that the segment is destroyed after we acquire the lock, then we don't release the lock. Does that matter?

Reply:
I don't know what the underlying impact would be since the lock is reentrant lock but I have fixed it.